core/src/main/java/io/questdb/griffin/engine/functions/date/TimestampAddFunctionFactory.java

```diff
@@ -51,125 +51,127 @@ public String getSignature() {
     }
 
     @Override
-    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) {
-
-        Function period = args.getQuick(0);
-        Function interval = args.getQuick(1);
-        if (period.isConstant()) {
-            char periodValue = period.getChar(null);
-            if (periodValue < addFunctionsMax) {
-                LongAddIntFunction func = addFunctions.getQuick(periodValue);
-                if (func != null) {
-                    if (interval.isConstant()) {
-                        if (interval.getInt(null) != Numbers.INT_NULL) {
-                            return new AddLongIntVarConstFunction(args.getQuick(2), interval.getInt(null), func, periodValue);
-                        }
-                        return TimestampConstant.NULL;
+    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        Function periodFunc = args.getQuick(0);
+        Function strideFunc = args.getQuick(1);
+        Function timestampFunc = args.getQuick(2);
+        LongAddIntFunction periodAddFunc;
+        char period;
+        int stride;
+
+        if (periodFunc.isConstant()) {
+            if ((period = periodFunc.getChar(null)) < addFunctionsMax && (periodAddFunc = addFunctions.getQuick(period)) != null) {
+                if (strideFunc.isConstant()) {
+                    if ((stride = strideFunc.getInt(null)) != Numbers.INT_NULL) {
+                        return new TimestampAddConstConstVar(period, periodAddFunc, stride, timestampFunc);
+                    } else {
+                        throw SqlException.$(argPositions.getQuick(1), "`null` is not a valid stride");
                     }
-                    return new AddLongIntVarVarFunction(args.getQuick(2), args.getQuick(1), func, periodValue);
                 }
+                return new TimestampAddConstVarVar(period, periodAddFunc, strideFunc, timestampFunc);
+            } else {
+                throw SqlException.$(argPositions.getQuick(0), "invalid time period unit '" + period + "'");
```

Comment:
SqlException has facility to avoid string concatenation cc @puzpuzpuz  Also 0 error position is a insufficient effort